src/bin/pg_autoctl/pgsql.c

```diff
@@ -227,35 +438,87 @@ pgsql_open_connection(PGSQL *pgsql)
  * when it could connect, false otherwise.
  */
 static PGconn *
-pgsql_retry_open_connection(PGSQL *pgsql)
+pgsql_retry_open_connection(PGSQL *pgsql, uint64_t startTime)
 {
 	PGconn *connection = NULL;
-	int attempts = 0;
-	bool retry = true;
 	bool connectionOk = false;
-	uint64_t startTime = time(NULL);
+
+	PGPing lastWarningMessage = PQPING_OK;
+	uint64_t lastWarningTime = 0;
 
 	log_warn("Failed to connect to \"%s\", retrying until "
 			 "the server is ready", pgsql->connectionString);
 
-	while (retry)
+	/* should not happen */
+	if (pgsql->retryPolicy.maxR == 0)
 	{
+		return NULL;
+	}
+
+	/* reset our internal counter before entering the retry loop */
+	pgsql->retryPolicy.attempts = 1;
+
+	while (!connectionOk)
+	{
+		int sleep = 0;
 		uint64_t now = time(NULL);
 
-		if ((now - startTime) >= POSTGRES_PING_RETRY_TIMEOUT)
+		/* Any signal is reason enough to break out from this retry loop. */
+		if (asked_to_quit ||
+			asked_to_stop ||
+			asked_to_stop_fast ||
+			asked_to_reload)
+		{
+			break;
+		}
+
+		/*
+		 * We stop retrying as soon as we have spent all of our time budget or
+		 * all of our attempts count budget, whichever comes first.
+		 *
+		 * maxR = 0 (zero) means no retry at all, checked before the loop
+		 * maxR < 0 (zero) means unlimited number of retries
+		 */
+		if ((now - startTime) >= pgsql->retryPolicy.maxT ||
+			(pgsql->retryPolicy.maxR > 0 &&
+			 pgsql->retryPolicy.attempts >= pgsql->retryPolicy.maxR))
 		{
-			log_warn("Failed to connect to \"%s\" after %d attempts, "
-					 "stopping now", pgsql->connectionString, attempts);
+			(void) log_connection_error(connection, LOG_WARN);
+
+			log_error("Failed to connect to \"%s\" "
+					  "after %d attempts in %d seconds, "
+					  "pg_autoctl stops retrying now",
+					  pgsql->connectionString,
+					  pgsql->retryPolicy.attempts,
+					  (int) (now - startTime));
 			break;
 		}
 
-		++attempts;
+		/*
+		 * Now compute how much time to wait for this round, and increment how
+		 * many times we tried to connect already.
+		 */
+		sleep = pgsql_compute_connection_retry_sleep_time(pgsql);
+
+		/* we have milliseconds, pg_usleep() wants microseconds */
+		(void) pg_usleep(sleep * 1000);
+
+		log_debug("PQping(%s): slept %d ms on attempt %d",
+				  pgsql->connectionString,
+				  pgsql->retryPolicy.sleepTime,
+				  pgsql->retryPolicy.attempts);
+
 		switch (PQping(pgsql->connectionString))
```

Comment:
After doing PQping we `now` should be calculated again, otherwise the below log messages use the now from before connecting. Another option is to not use `now` in those error mesasges, but use `time(NULL)` directly.

Reply:
Done!